tests/circuit_graph/test_qasm.py

```diff
@@ -0,0 +1,719 @@
+    when circuits are created via QNodes."""
+
+    def test_empty_circuit(self):
+        """Test that an empty QNode is properly
+        serialized into an empty QASM program."""
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def qnode():
+            return qml.expval(qml.PauliZ(0))
+
+        # construct the qnode circuit
+        qnode()
+
+        res = qnode.circuit.to_qasm()
+        expected = dedent(
+            """\
+            OPENQASM 2.0;
+            include "qelib1.inc";
+            qreg q[1];
+            creg c[1];
+            measure q[0] -> c[0];
+            """
+        )
+        assert res == expected
+
+    def test_native_qasm_gates(self):
+        """Test that a QNode containing solely native QASM
+        gates is properly serialized."""
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def qnode():
+            qml.RX(0.43, wires=0)
+            qml.RY(0.35, wires=1)
+            qml.RZ(0.35, wires=2)
+            qml.CNOT(wires=[0, 1])
+            qml.Hadamard(wires=2)
+            qml.CNOT(wires=[2, 0])
+            qml.PauliX(wires=1)
+            return qml.expval(qml.PauliZ(0))
+
+        # construct the qnode circuit
+        qnode()
+        res = qnode.circuit.to_qasm()
+
+        expected = dedent(
+            """\
+            OPENQASM 2.0;
+            include "qelib1.inc";
+            qreg q[3];
+            creg c[3];
+            rx(0.43) q[0];
+            ry(0.35) q[1];
+            rz(0.35) q[2];
+            cx q[0],q[1];
+            h q[2];
+            cx q[2],q[0];
+            x q[1];
+            measure q[0] -> c[0];
+            measure q[1] -> c[1];
+            measure q[2] -> c[2];
+            """
+        )
+
+        assert res == expected
+
+    def test_parametrized_native_qasm_gates(self):
+        """Test that a QNode containing solely native QASM
+        gates, as well as input parameters, is properly serialized.
+        In addition, double check the serialization changes as parameters
+        are changed."""
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def qnode(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y[0], wires=1)
+            qml.RZ(y[1], wires=2)
+            qml.CNOT(wires=[0, 1])
+            qml.Hadamard(wires=2)
+            qml.CNOT(wires=[2, 0])
+            qml.PauliX(wires=1)
+            return qml.expval(qml.PauliZ(0))
+
+        expected = dedent(
+            """\
+            OPENQASM 2.0;
+            include "qelib1.inc";
+            qreg q[3];
+            creg c[3];
+            rx({}) q[0];
+            ry({}) q[1];
+            rz({}) q[2];
+            cx q[0],q[1];
+            h q[2];
+            cx q[2],q[0];
+            x q[1];
+            measure q[0] -> c[0];
+            measure q[1] -> c[1];
+            measure q[2] -> c[2];
+            """

```

Comment:
Got a bit confused here, thought for a moment this was valid qasm spec :laughing:

Reply:
Might be worth codifying here directly what the actual expected strings are, since there's a bit much to digest in the `assert` statement (`format`, `hstack`, `*`, etc)

Reply:
Good call, I've split this out into separated hard-coded strings.